### ISSUE TITLE
Zookeeper connection number continuously increased after posting /refresh endpoint

### ISSUE THREAD

**Issue by wuxianInfi** [spring-cloud/spring-cloud-zookeeper#123]
**Steps to Reproduce**：
1. Start an application with spring-cloud-zookeeper-config
2. Type "netstat -Antlu -p tcp | grep '\.2181' | wc -l" in terminal to see zookeeper connection numbers (e.g. 1)
3. Post /refresh endpoint for several times
4. Type "netstat -Antlu -p tcp | grep '\.2181' | wc -l" again in terminal to see zookeeper connection numbers, now it comes 6

**Analysis Result**:
ZookeeperConfigBootstrapConfiguration import ZookeeperAutoConfiguration, which means zookeeper CuratorFramework is initialized during "bootstarp" phase. After post /refresh, Spring Cloud ContextRefresher is invoked. It seems will refresh all the beans initialized during "bootsrap" phase including CuratorFramework. But due to some reason, the previous one is not destroyed correctly (destoryMethod not invoked).

**Question**:
1. Why the previous CuratorFramework is not destoryed correctly, is it a bug?
2. Is there a way to disable re-initialize CuratorFramework after invoke ContextRefresher. It seems no needed and waste resources.

**Comment by wuxianInfi** [comment]
Any updates about this? Help wanted.

**Comment by dsyer** [comment]
I don't see the behaviour you describe with the `spring-cloud-zookeeper-sample` in this repo. Maybe you are using an older version or something?

**Comment by wuxianInfi** [comment]
[boot_server2.zip](https://github.com/spring-cloud/spring-cloud-zookeeper/files/982619/boot_server2.zip)

I upload a sample which can reproduce the problem, please have a look.

**Comment by dsyer** [comment]
I can't compile that project (loads of unresolved dependencies). If I remove the custom repository I can run it, but I can't see any issues as described above. Maybe you are doing something else that I'm not seeing?

```
$ sudo netstat -ntlu -p tcp | grep '.2181' | wc -l
1
$ curl localhost:8082/refresh -d {}
$ curl localhost:8082/refresh -d {}
$ curl localhost:8082/refresh -d {}
$ curl localhost:8082/refresh -d {}
$ sudo netstat -ntlu -p tcp | grep '.2181' | wc -l
1
```

The `netstat` command you listed is also invalid on my system. Maybe that's the difference?

**Comment by wuxianInfi** [comment]
I tried the same netstat command, also has the problem:
![image](https://cloud.githubusercontent.com/assets/13640013/25801546/a71fd27e-3420-11e7-827d-1dc8ff35d55d.png)

refresh endpoint should return a string array. and refresh is invoked, you should see following logs:
![image](https://cloud.githubusercontent.com/assets/13640013/25801581/dbf45132-3420-11e7-94b2-10248c88e318.png)

**Comment by dsyer** [comment]
Well, what can I say. It works fro me. Your custom repository could be the problem. Did you try without that?

What is your OS/Java version etc.?

Please don't use screenshots for documenting issues.

**Comment by wuxianInfi** [comment]
I'm not sure whether it depends on the zookeeper version. My version is 3.4.9, what's yours?
Can you upload your project after custom repository been removed? Let me have a try with this.

**Comment by dsyer** [comment]
I just used the latest zookeeper from dockerhub (3.4.10 I think). But that has nothing to do with /refresh. The code I was running is here: https://github.com/scratches/zookeeper-sample.

**Comment by ma3a** [comment]
Hi @dsyer, 
I could confirm that this issue exists. 
We've faced it on our testing environment, when we update all application specific properties in ZK. We've see a big thread grow: more than 300 thread in 2 minutes, all thread with `Curator` prefix. Also i could confirm when `ContextRefresher` is invoked, destroy method `close`
```java 
@Bean(destroyMethod = "close")
	@ConditionalOnMissingBean
	public CuratorFramework curatorFramework(RetryPolicy retryPolicy, ZookeeperProperties properties) throws Exception {
		CuratorFrameworkFactory.Builder builder = CuratorFrameworkFactory.builder();
```
is never called. 
Here is a project that reproduce this issue
https://github.com/ma3a/zookeeperthreads

**Comment by spencergibb** [comment]
@ma3a this has been fixed via https://github.com/spring-cloud/spring-cloud-commons/pull/267. You should be able to try Dalston or Edgware BUILD-SNAPSHOTS.

**Comment by ma3a** [comment]
@spencergibb, test on `Edgware.BUILD-SNAPSHOT` and problem is fixed, `CuratorFramework.close()` invoked successfuly

**Comment by spencergibb** [comment]
Closed via  spring-cloud/spring-cloud-commons#267
